starlite/handlers/websocket_handlers.py

```diff
@@ -94,3 +100,144 @@ def _validate_handler_function(self) -> None:
+        *,
+        dependencies: Dependencies | None = None,
+        exception_handlers: dict[int | type[Exception], ExceptionHandler] | None = None,
+        guards: list[Guard] | None = None,
+        middleware: list[Middleware] | None = None,
+        name: str | None = None,
+        opt: dict[str, Any] | None = None,
+        signature_namespace: Mapping[str, Any] | None = None,
+        mode: Literal["text", "binary"] = "text",
+        on_accept: Callable[[WebSocket], SyncOrAsyncUnion[None]] | None = None,
+        on_disconnect: Callable[[WebSocket], SyncOrAsyncUnion[None]] | None = None,
+        **kwargs: Any,
+    ) -> None:
+        self._mode = mode
+        self._pass_socket = False
+        self._on_accept = AsyncCallable(on_accept) if on_accept else None
+        self._on_disconnect = AsyncCallable(on_disconnect) if on_disconnect else None
+
+        super().__init__(
+            path=path,
+            dependencies=dependencies,
+            exception_handlers=exception_handlers,
+            guards=guards,
+            middleware=middleware,
+            name=name,
+            opt=opt,
+            signature_namespace=signature_namespace,
+            **kwargs,
+        )
+
+    def __call__(self, listener_callback: AnyCallable) -> websocket_listener:
+        self._validate_listener_callback(listener_callback)
+
+        listener_callback_signature = inspect.signature(listener_callback)
+
+        if not is_async_callable(listener_callback):
+            listener_callback = AsyncCallable(listener_callback)
+
+        should_receive_json = listener_callback_signature.parameters["data"].annotation not in {
+            "str",
+            str,
+            "bytes",
+            bytes,
+        }
+
+        async def listener_fn(socket: WebSocket, **kwargs: Any) -> None:
+            await socket.accept()
+            if self._on_accept:
+                await self._on_accept(socket)
+            if self._pass_socket:
+                kwargs["socket"] = socket
+            while True:
+                try:
+                    received_data = await socket.receive_data(mode=self._mode)  # pyright: ignore
+                    if should_receive_json:
+                        received_data = decode_json(received_data)
+                    data_to_send = await listener_callback(data=received_data, **kwargs)
+                    if isinstance(data_to_send, str):
+                        await socket.send_text(data_to_send)
+                    elif isinstance(data_to_send, bytes):
+                        await socket.send_bytes(data_to_send)
+                    else:
+                        await socket.send_json(data_to_send)
+                except WebSocketDisconnect:
+                    if self._on_disconnect:
+                        await self._on_disconnect(socket)
+                    break
+
+        # make our listener_fn look like the callback, so we get a correct signature model
```

Comment:
Because no callable with the correct signature exists. 

There is the callback, which has the additional dependencies and `data` but not necessarily `socket`, and the handler function we create, which has `socket` but none of the additional dependencies defined by the callback.

So we need to create a signature from a function that looks exactly like the callback, minus `data`, and which is guaranteed to include `socket`.

Maybe I don't follow what you mean, but I don't think your suggestion would accomplish that?

Reply:
Ok, lets revisit this code again in the future for refactoring - i think we can address this in a more cohesive way with some of the approaches that have been recently introduced by @peterschutt, but its not a blocker.